README.md
# ESP-Engine README

Addressing the aggressive use of terminal based commands required for working with ESP-IDF (and other variants such as ADF etc.) this extension is developed to create an command like interface (with Key-Binding support) to easen the development.

>NOTE: This extension in the curernt version does not include the toolchain or ESP-IDF component repository, future updates will resolve this too.

## Features

The ESP-Engine contains following Commands:

* `ESP-Engine.Initiate`: Convert current working directory to template IDF project.
* `ESP-Engine.Build`: Compile the whole project and generate Bin files for flashing.
* `ESP-Engine.Clean`: Clean built files.
* `ESP-Engine.Flash`: FLashes the built binaries to specified target device.
* `ESP-Engine.Monitor`: Serial Monitor for Debugging.
* `ESP-Engine.Menuconfig`: Project Configuration window.
* `ESP-Engine.Upcoming`: And many upcoming feautres in the pipeline.

## Requirements

You are required to have:

1. ESP-Toolchain : Refer ESP-IDF docs @ <https://docs.espressif.com/projects/esp-idf/en/stable/get-started-cmake/index.html>.
2. ESP-IDF       : Use git to clone ESP-IDF (preferrably stable version) using ->

```git
git clone -b v3.2.2 --recursive https://github.com/espressif/esp-idf.git
```

## Known Issues

No major known issues, kindly post issues you find at the github Repository.

### Beta Release v0.0.1

Initial release of ESP-Engine.
Contains all basic commands ESP-IDF development (namely Build, Flash, Clean, Monitor, Menuconfig).
Support automatic conversion of root folder to ESP-IDF Template.

-----------------------------------------------------------------------------------------------------------


CHANGELOG.md
# Change Log

All notable changes to the "ESP-Engine" extension will be documented in this file.

## Unreleased

- Unreleased features include automatic toolchain setup and ESP-IDF cloning. These features are in pipeline for future updates.

## [Beta Release v0.0.1]

- Initial release, Contains basic features such as,

1. Build System Commands:
    * Build command
    * Flash Command
    * Clean Command
    * Monitor Command
    * Menuconfig Command
2. Automatic Root folder conversion to ESP-IDF Template (Requires Internet Connection).


src/esp-engine.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('ESP-IDF engine is now active!');
	let activity_status = true;
	let engine_terminal; // For unified terminal experience.
	let template_uri = "'https://github.com/espressif/esp-idf-template/archive/master.zip'";
	let template_location = '$loc\\esp-engine\\template.zip';

	// Initiate Command
	// Not completely implemented
	let initiate_command = vscode.commands.registerCommand('esp-engine.initiate', function (){
		vscode.window.showInformationMessage('Initiating Directory as ESP-IDF project ...');
		engine_terminal = vscode.window.createTerminal('esp-Init');
		if(vscode.workspace.findFiles('CMakeLists.txt')){
			// engine_terminal.sendText('$loc = Get-location', true);
			// engine_terminal.sendText('Remove-Item $loc/* -Force -Recurse', true);
			engine_terminal.sendText('mkdir /esp-engine/', true);
			engine_terminal.sendText('Invoke-WebRequest -Uri ' + template_uri + " -Outfile " + template_location, true);
			engine_terminal.sendText('Expand-Archive ' + template_location + " $loc\\esp-engine\\temp\\", true);
			engine_terminal.sendText('mv ' + " $loc\\esp-engine\\temp\\esp-idf-template-master\\* $loc\\", true);
			engine_terminal.sendText('Remove-Item $loc\\esp-engine\\ -Force -Recurse', true);
			vscode.window.showInformationMessage("Done.");
		}else{
			vscode.window.showInformationMessage("Not an Empty Folder!");
		}
	});
	context.subscriptions.push(initiate_command);

	// Build Command
	let build_command = vscode.commands.registerCommand('esp-engine.build', function(){
		if(activity_status){
			console.log("'CMakeLists.txt' file is present, initiating build.");
			engine_terminal = vscode.window.createTerminal("esp-build");
			vscode.workspace.saveAll();
			vscode.window.showInformationMessage("Project Compilation Initiated.");
			engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py build", true);
			engine_terminal.show();
		}
	});
	context.subscriptions.push(build_command);

	// Menuconfig Command
	let menuconfig_command = vscode.commands.registerCommand('esp-engine.configure', function(){
		if(activity_status){
			engine_terminal = vscode.window.createTerminal("esp-config");
			vscode.window.showInformationMessage("Terminal size should be atleast 19 Lines by 80 Columns");
			engine_terminal.show();
			engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py menuconfig", true);
		}
	});
	context.subscriptions.push(menuconfig_command);

	// Clean Command
	let clean_command = vscode.commands.registerCommand('esp-engine.clean', function(){
		if(activity_status){
			engine_terminal = vscode.window.createTerminal("esp-clean");
			vscode.window.showInformationMessage("Cleaning built files.");
			engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py clean", true);
			engine_terminal.dispose();
			vscode.window.showInformationMessage("We've cleaned the built files.s");
		}
	});
	context.subscriptions.push(clean_command);

	// Flash command
	let flash_command = vscode.commands.registerCommand('esp-engine.flash', function(){
		if(activity_status){
			engine_terminal = vscode.window.createTerminal("esp-flash");
			vscode.window.showInformationMessage("Flashing Target", );
			engine_terminal.show();
			engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py flash", true);
		}
	});
	context.subscriptions.push(flash_command);

	// Monitor command
	let monitor_command = vscode.commands.registerCommand('esp-engine.monitor', function(){
		if(activity_status){
			engine_terminal = vscode.window.createTerminal("esp-monitor");
			vscode.window.showInformationMessage("Opening Serial Monitor");
			engine_terminal.show();
			engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py monitor", true);
		}
	});
	context.subscriptions.push(monitor_command);

	// Setup Command
	// let setup_command = vscode.commands.registerCommand('esp-engine.monitor', function(){
	// 	if(activity_status){
	// 		engine_terminal = vscode.window.createTerminal("esp-monitor");
	// 		vscode.window.showInformationMessage("Opening Serial Monitor");
	// 		engine_terminal.show();
	// 		engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py monitor", true);
	// 	}
	// });
	// context.subscriptions.push(setup_command);
}
exports.activate = activate;

// this method is called when your extension is deactivated
function deactivate() {}

module.exports = {
	activate,
	deactivate
};


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('ESP-IDF engine is now active!');
	let activity_status = true;
	let engine_terminal = vscode.window.createTerminal("esp-engine");
	let template_uri = "'https://github.com/espressif/esp-idf-template/archive/master.zip'";

	// Initiate Command
	// Not completely implemented
	let initiate_command = vscode.commands.registerCommand('esp-engine.initiate', async function (){
		// Init Message
		vscode.window.showInformationMessage('Initiating Directory as ESP-IDF project ... It will take a few seconds.');
		// Setting up CWD as a ESP-IDF template.
		// engine_terminal = vscode.window.createTerminal('esp-Init');
		await engine_terminal.sendText("$loc = Get-Location", true);
		await engine_terminal.sendText("mkdir $loc\\.esp-engine", true);
		await engine_terminal.sendText("Invoke-WebRequest -Uri " + template_uri + " -OutFile $loc\\.esp-engine\\template.zip", true);
		await engine_terminal.sendText("Expand-Archive $loc\\.esp-engine\\template.zip $loc\\.esp-engine\\", true);
		await engine_terminal.sendText("mv $loc\\.esp-engine\\esp-idf-template-master\\* $loc\\", true);
		await engine_terminal.sendText("Remove-Item $loc\\.esp-engine\\esp-idf-template-master -Force -Recurse");
	});
	context.subscriptions.push(initiate_command);

	// Build Command
	let build_command = vscode.commands.registerCommand('esp-engine.build', function(){
		if(activity_status){
			console.log("'CMakeLists.txt' file is present, initiating build.");
			// engine_terminal = vscode.window.createTerminal("esp-build");
			vscode.workspace.saveAll();
			engine_terminal.show();
			vscode.window.showInformationMessage("Project Compilation Initiated.");
			engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py build", true);
		}
	});
	context.subscriptions.push(build_command);

	// Menuconfig Command
	let menuconfig_command = vscode.commands.registerCommand('esp-engine.config', function(){
		if(activity_status){
			// engine_terminal = vscode.window.createTerminal("esp-config");
			engine_terminal.show();
			vscode.window.showInformationMessage("Terminal size should be atleast 19 Lines by 80 Columns");
			engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py menuconfig", true);
			
		}
	});
	context.subscriptions.push(menuconfig_command);

	// Clean Command
	let clean_command = vscode.commands.registerCommand('esp-engine.clean', async function(){
		if(activity_status){
			// engine_terminal = vscode.window.createTerminal("esp-clean");
			vscode.window.showInformationMessage("Cleaning built files.");
			await engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py clean", true);
			vscode.window.showInformationMessage("We've cleaned the built files.");
		}
	});
	context.subscriptions.push(clean_command);

	// Flash command
	let flash_command = vscode.commands.registerCommand('esp-engine.flash', async function(){
		if(activity_status){
			// engine_terminal = vscode.window.createTerminal("esp-flash");
			vscode.window.showInformationMessage("Flashing Target", );
			let baudrate = await vscode.window.showInputBox({
				prompt: "Specify Baudrate.",
				value: "115200"
			});
			let port = await vscode.window.showInputBox({
				prompt: "Specify Port Address.",
				value: "COM10"
			});
			await engine_terminal.show();
			await engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py -p " + port + " -b " + baudrate + " flash", true);
		}
	});
	context.subscriptions.push(flash_command);

	// Monitor command
	let monitor_command = vscode.commands.registerCommand('esp-engine.monitor', function(){
		if(activity_status){
			// engine_terminal = vscode.window.createTerminal("esp-monitor");
			vscode.window.showInformationMessage("Opening Serial Monitor");
			engine_terminal.show();
			engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py monitor", true);
		}
	});
	context.subscriptions.push(monitor_command);

	// Setup Command
	// let setup_command = vscode.commands.registerCommand('esp-engine.monitor', function(){
	// 	if(activity_status){
	// 		engine_terminal = vscode.window.createTerminal("esp-monitor");
	// 		vscode.window.showInformationMessage("Opening Serial Monitor");
	// 		engine_terminal.show();
	// 		engine_terminal.sendText("python $env:IDF_PATH/tools/idf.py monitor", true);
	// 	}
	// });
	// context.subscriptions.push(setup_command);
}
exports.activate = activate;

// this method is called when your extension is deactivated
function deactivate() {}

module.exports = {
	activate,
	deactivate
};
